File: src/test/java/net/syscon/elite/util/JwtAuthenticationHelper.java
package net.syscon.elite.util;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import org.apache.commons.codec.binary.Base64;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.InputStream;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyStore;
import java.security.PublicKey;
import java.security.interfaces.RSAPrivateCrtKey;
import java.security.spec.RSAPublicKeySpec;
import java.util.Date;
import java.util.HashMap;
import java.util.UUID;

@Component
public class JwtAuthenticationHelper {
    private final KeyPair keyPair;

    public JwtAuthenticationHelper(@Value("${jwt.signing.key.pair}") final String privateKeyPair,
                                   @Value("${jwt.keystore.password}") final String keystorePassword,
                                   @Value("${jwt.keystore.alias:elite2api}") final String keystoreAlias) {
        keyPair = getKeyPair(new ByteArrayResource(Base64.decodeBase64(privateKeyPair)), keystoreAlias, keystorePassword.toCharArray());
    }

    public String createJwt(final JwtParameters parameters) {

        final var claims = new HashMap<String, Object>();


        if (parameters.getUsername() != null) {
            claims.put("user_name", parameters.getUsername());
        }
        claims.put("client_id", parameters.getClientId());
        claims.put("internalUser", parameters.isInternalUser());

        if (parameters.getRoles() != null && !parameters.getRoles().isEmpty())
            claims.put("authorities", parameters.getRoles());

        if (parameters.getScope() != null && !parameters.getScope().isEmpty())
            claims.put("scope", parameters.getScope());

        return Jwts.builder()
                .setId(UUID.randomUUID().toString())
                .setSubject(parameters.getUsername())
                .addClaims(claims)
                .setExpiration(new Date(System.currentTimeMillis() + parameters.getExpiryTime().toMillis()))
                .signWith(SignatureAlgorithm.RS256, keyPair.getPrivate())
                .setIssuer("http://localhost:8080/auth/issuer")
                .setHeaderParam("typ", "JWT")
                .compact();
    }

    private KeyPair getKeyPair(final Resource resource, final String alias, final char[] password) {
        InputStream inputStream = null;
        KeyStore store;
        try {
            store = KeyStore.getInstance("jks");
            inputStream = resource.getInputStream();
            store.load(inputStream, password);
            final RSAPrivateCrtKey key = (RSAPrivateCrtKey) store.getKey(alias, password);
            final RSAPublicKeySpec spec = new RSAPublicKeySpec(key.getModulus(), key.getPublicExponent());
            final PublicKey publicKey = KeyFactory.getInstance("RSA").generatePublic(spec);
            return new KeyPair(publicKey, key);
        }
        catch (Exception e) {
            throw new IllegalStateException("Cannot load keys from store: " + resource, e);
        }
        finally {
            try {
                if (inputStream != null) {
                    inputStream.close();
                }
            }
            catch (IOException e) {
                throw new IllegalStateException("Cannot close open stream: ", e);
            }
        }
    }


}

File: src/test/java/uk/gov/justice/hmpps/prison/util/JwtAuthenticationHelper.kt
package uk.gov.justice.hmpps.prison.util

import io.jsonwebtoken.Jwts
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.security.oauth2.jwt.JwtDecoder
import org.springframework.security.oauth2.jwt.NimbusJwtDecoder
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.interfaces.RSAPublicKey
import java.time.Duration
import java.util.Date
import java.util.UUID

@Configuration
class JwtAuthenticationHelper {
  private val keyPair: KeyPair = KeyPairGenerator.getInstance("RSA").apply { initialize(2048) }.generateKeyPair()

  @Bean
  @Primary
  fun jwtDecoder(): JwtDecoder = NimbusJwtDecoder.withPublicKey(keyPair.public as RSAPublicKey).build()

  fun createJwt(parameters: JwtParameters): String = with(parameters) {
    createJwt(
      username = username,
      scope = scope,
      grantType = grantType,
      roles = roles,
      expiryTime = expiryTime,
      clientId = clientId,
      internalUser = internalUser,
    )
  }

  fun createJwt(
    username: String? = null,
    scope: List<String>? = null,
    grantType: String? = null,
    roles: List<String>? = null,
    expiryTime: Duration = Duration.ofDays(1),
    clientId: String = "prison-api-client",
    internalUser: Boolean = true,
  ): String {
    val claims = mutableMapOf<String, Any?>(
      "client_id" to clientId,
      "internalUser" to internalUser,
      "sub" to (username ?: clientId),
    ).apply {
      username?.let { this["user_name"] = username }
      roles?.let {
        // ensure that all roles have a ROLE_ prefix
        this["authorities"] = roles.map { "ROLE_${it.substringAfter("ROLE_")}" }
      }
      scope?.let { this["scope"] = scope }
      grantType?.let { this["grant_type"] = grantType }
    }
    return Jwts.builder()
      .id(UUID.randomUUID().toString())
      .claims(claims)
      .expiration(Date(System.currentTimeMillis() + expiryTime.toMillis()))
      .signWith(keyPair.private, Jwts.SIG.RS256)
      .header().add("typ", "JWT").add("kid", "dps-client-key").and()
      .compact()
  }
}
